### PR TITLE
Add surface curvature (H, K) as kNN-estimated input features for tau

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Surface curvature (mean curvature H and Gaussian curvature K) encodes local geometry that is physically meaningful for boundary-layer behavior and wall shear stress. Concave vs convex surface regions have distinct pressure gradient structures — mean curvature H is related to the normal pressure gradient via the Young–Laplace relation, while Gaussian curvature K distinguishes saddle regions (K<0) from dome/bowl regions (K>0). Providing these as explicit input features should give the model physics-grounded geometry priors that STRING-separable RFF encoding cannot capture, and may specifically reduce the tau_y and tau_z errors (currently 8.3631% and 9.8099% val, 2.3–2.7× above AB-UPT targets of 3.65% and 3.63%).

This experiment was originally assigned as PR #605 (edward/surface-curvature-features) but that PR was auto-merged as an empty diff before the student ran any experiments. This is the clean re-run with updated SOTA baseline.

## Instructions

You need to (1) compute surface curvature features from point coordinates, (2) concatenate them to the surface input features, and (3) update the input projection accordingly. All changes go in `target/train.py`.

### Step 1 — Compute mean H and Gaussian K from k-NN

Add this function near the top of train.py (after imports):

```python
def compute_surface_curvature(surface_pts: torch.Tensor, k: int = 20) -> torch.Tensor:
    """
    Estimate per-point mean curvature H and Gaussian curvature K via PCA on k-NN patches.

    Args:
        surface_pts: (N, 3) XYZ coordinates
        k: number of nearest neighbors for local patch

    Returns:
        (N, 2) tensor [H, K], zero-mean / unit-std, clipped at ±3σ
    """
    N = surface_pts.shape[0]
    device = surface_pts.device

    # Build k-NN (brute-force; N is typically 65k — fast enough on GPU)
    # Pairwise distances
    diff = surface_pts.unsqueeze(1) - surface_pts.unsqueeze(0)  # (N,N,3)
    dist2 = (diff ** 2).sum(-1)                                  # (N,N)
    # k+1 because self is included
    _, idx = dist2.topk(k + 1, dim=1, largest=False)            # (N, k+1)
    neighbors = surface_pts[idx[:, 1:]]                          # (N, k, 3)

    # Center patches
    center = surface_pts.unsqueeze(1)                            # (N,1,3)
    local = neighbors - center                                   # (N,k,3)

    # 3×3 covariance → PCA
    cov = torch.bmm(local.transpose(1, 2), local) / k           # (N,3,3)
    # eigenvalues ascending: λ₀ (normal direction), λ₁ λ₂ (tangent plane)
    eigvals, eigvecs = torch.linalg.eigh(cov)                   # (N,3), (N,3,3)
    normal = eigvecs[:, :, 0]                                    # (N,3) surface normal

    # Project neighbors onto tangent plane
    proj = local - (local * normal.unsqueeze(1)).sum(-1, keepdim=True) * normal.unsqueeze(1)
    # u, v in tangent plane (use eigvecs[:,1] and eigvecs[:,2])
    u_ax = eigvecs[:, :, 1]   # (N,3)
    v_ax = eigvecs[:, :, 2]   # (N,3)
    pu = (proj * u_ax.unsqueeze(1)).sum(-1)   # (N,k)
    pv = (proj * v_ax.unsqueeze(1)).sum(-1)   # (N,k)
    ph = (local * normal.unsqueeze(1)).sum(-1) # (N,k) heights

    # Fit quadratic form: h ≈ a*u² + b*u*v + c*v²  (linear + const absorbed into residual)
    # Design matrix columns: [u², u*v, v², u, v, 1]
    A = torch.stack([pu**2, pu*pv, pv**2, pu, pv, torch.ones_like(pu)], dim=2)  # (N,k,6)
    # Least-squares solve: (AᵀA)⁻¹Aᵀ h
    AtA = torch.bmm(A.transpose(1, 2), A)           # (N,6,6)
    Ath = torch.bmm(A.transpose(1, 2), ph.unsqueeze(2))  # (N,6,1)
    try:
        coeffs = torch.linalg.solve(AtA, Ath).squeeze(-1)  # (N,6)
    except Exception:
        coeffs = torch.zeros(N, 6, device=device)

    # Shape operator from quadratic coefficients [a, b, c, ...]
    a, b, c = coeffs[:, 0], coeffs[:, 1], coeffs[:, 2]
    # Principal curvatures via eigenvalues of [[2a, b],[b, 2c]]
    trace = 2 * a + 2 * c
    det = 4 * a * c - b ** 2
    disc = torch.clamp((trace / 2) ** 2 - det, min=0.0)
    sqrt_disc = disc.sqrt()
    k1 = trace / 2 + sqrt_disc
    k2 = trace / 2 - sqrt_disc
    H = (k1 + k2) / 2   # mean curvature
    K = k1 * k2          # Gaussian curvature

    feats = torch.stack([H, K], dim=1)  # (N,2)

    # Clip at ±3σ then normalize to zero-mean / unit-std
    for i in range(2):
        mu = feats[:, i].mean()
        sigma = feats[:, i].std().clamp(min=1e-6)
        feats[:, i] = (feats[:, i] - mu) / sigma
        feats[:, i] = feats[:, i].clamp(-3.0, 3.0)

    return feats  # (N,2)
```

**Memory note**: For N=65536 the `(N,N)` pairwise distance matrix is 65536²×4 bytes ≈ 16 GB, which is too large. Use a chunked approach or use `torch.cdist` with chunk size ≤8192 rows:

```python
# Chunked k-NN to avoid OOM
chunk = 8192
idx_list = []
for start in range(0, N, chunk):
    end = min(start + chunk, N)
    d2 = torch.cdist(surface_pts[start:end], surface_pts)  # (chunk, N)
    _, ki = d2.topk(k + 1, dim=1, largest=False)
    idx_list.append(ki[:, 1:])  # drop self
idx = torch.cat(idx_list, dim=0)  # (N, k)
```

Replace the pairwise-distance block above with this chunked version.

### Step 2 — Call it during data loading / forward pass

The simplest injection point is immediately before the surface features are passed into the model. In the training loop where surface points are prepared (look for `surface_pts` or `surf_x`), add:

```python
with torch.no_grad():
    curv = compute_surface_curvature(surface_pts.float(), k=20)  # (N,2)
# Concatenate as additional channels: [x, y, z, ..., H, K]
surface_feats = torch.cat([surface_feats, curv], dim=-1)
```

For efficiency, precompute curvature once per mesh (not per mini-batch) and cache it. If the dataset class exposes a per-case cache directory, save as `curvature_{case_id}.pt` and load on subsequent passes.

### Step 3 — Update input projection dimension

Find where the model input `in_channels` or `in_features` is set for surface points and add 2:

```python
# Before: in_features = 3 + <other_surface_features>
# After:  in_features = 3 + <other_surface_features> + 2  # +H, +K
```

Check `train.py` for where the Transolver model is initialized and update accordingly.

### Run command

```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent thorfinn --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 13 --epochs 13 \
  --vol-points-schedule "0:16384:3:32768:6:49152:9:65536" \
  --no-compile-model \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --wandb-group "thorfinn-surface-curvature" \
  --wandb-run-name "thorfinn/curvature-H-K-knn20"
```

### Kill gates

- **EP2 (step 21,729)**: if val_abupt > 12.0% → kill immediately, post results and close PR.
- **EP3 (step 32,594)**: if val_abupt > 8.0% → kill, post results and close PR.
- **EP4 (step 38,030)**: report full per-channel metrics and mark PR ready for review.

## Baseline

Current single-model SOTA: **PR #592 (alphonse)** — L=5 Transolver

| Metric | Val | Test |
|---|---|---|
| **abupt (primary)** | **6.5985%** | 7.9915% |
| surface_pressure | 4.3322% | 4.8090% |
| volume_pressure | 3.9456% | 11.4652% |
| wall_shear_x (tau_x) | 6.5420% | 8.1226% |
| wall_shear_y (tau_y) | 8.3631% | 10.1740% |
| wall_shear_z (tau_z) | 9.8099% | 12.4058% |

**AB-UPT targets** (from paper): tau_y=3.65%, tau_z=3.63%, surface_pressure=3.82%, volume_pressure=6.08%

tau_y and tau_z have the largest remaining gaps to the AB-UPT floor — 2.3× and 2.7× respectively. Surface curvature features are specifically hypothesized to close this gap.

**Beat this to merge**: val_abupt < **6.5985%** (lower is better)

Baseline W&B run: `4k25s25e` — https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/4k25s25e

Reproduce command (baseline, no curvature):
```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent alphonse --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 13 --epochs 13 \
  --vol-points-schedule "0:16384:3:32768:6:49152:9:65536" \
  --no-compile-model \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536
```

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Status update — your pod is in watchdog hold state, waiting for the previous flow-aligned-tau run to finish. That zombie run is at step 36,417/~37,500 (~97% complete) and should finish within the next hour.

Once it finishes, the watchdog will auto-transition your pod to this PR's surface-curvature experiment. No action needed on your end.

**EP gate reminders for when you start:**
- EP1 (~step 10,864): Informational only — post results but no kill
- EP2 (~step 21,729): Kill threshold val_abupt > 12.0%

**Current baseline to beat:** val_primary/abupt_axis_mean_rel_l2_pct = **6.5985%** (single-model SOTA, alphonse PR #592)

**Comment by morganmcg1** [comment]
ADVISOR KILL VERDICT — PR #651 (thorfinn, q4z6tmmo, surface curvature features)

**EP2 Result: val_abupt = 14.613% at step ~21,729 → FAIL** (threshold: < 12.0%)

The run continued to completion but never recovered:
- EP2 (step ~21,729): val_abupt = 14.613% → **ABOVE kill threshold of 12.0%**
- Final (step 37,331): val_abupt = 12.487% → **Still above 12.0%**

The surface curvature features (H, K) as tau input features did not improve convergence and actually degraded it significantly versus baseline (~25-27% at EP1, not 14.6%). This approach is not viable in this form.

**This PR is being closed.** The experiment is terminated.

Thorfinn: you will receive a new assignment shortly. Thank you for running this experiment — the negative result is informative and rules out naive kNN-estimated curvature as a useful tau feature.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #651 because ADVISOR: Closing — EP2 val_abupt=14.613% exceeds kill threshold of 12.0%. Run completed at 12.487%, still above threshold. Surface curvature features (H, K) as tau inputs did not improve convergence. Negative result rules out this approach. Thorfinn will receive new assignment..
